src/language_features/references.rs

```diff
@@ -43,46 +44,73 @@ pub fn editor_references(
         ReferencesResponse::Array(locations) => Some(locations),
         ReferencesResponse::None => None,
     } {
+        // Sort locations by (filename, line)
+        let mut locations = locations.to_vec();
+
+        locations
+            .sort_unstable_by_key(|location| {
+              (location.uri.to_file_path(),
+              location.range.start.line)
+          });
+
         let content = locations
             .iter()
-            .map(|location| {
-                let p = location.range.start;
-                let path = location.uri.to_file_path().unwrap();
-                let filename = path.to_str().unwrap();
-                let file = File::open(filename);
+            .group_by(|location|{
+              location.uri.to_file_path()
+            })
+            .into_iter()
+            .map(|(filename, group)| {
+                let filename = filename.unwrap();
+                let name = filename.to_str().unwrap();
+                let file = File::open(name);
```

Comment:
`File::open` can take path as well, no need to convert it to string. Also we could lift making relative path string from line 87 because it will be the same each iteration of loop. I imagine something like that:

```
                let filename = filename.unwrap();
                let file = File::open(&filename);
                let name = filename
                    .strip_prefix(&ctx.root_path)
                    .ok()
                    .and_then(|p| Some(p.to_str().unwrap()))
                    .or_else(|| filename.to_str())
                    .unwrap();
```

Reply:
Oh yeah, of course. I blindly copy/pasted this part. I'll adapt it